File: hal/HAL/cpp/AnalogOutput.cpp
/*
 * AnalogOutput.cpp
 *
 *  Created on: Aug 27, 2017
 *      Author: root
 */

#include "AnalogOutput.h"
#include "Analog.h"
#include "portexception.h"

namespace flashlib {

namespace hal {

AnalogOutput::AnalogOutput(uint8_t port){
	if(!HAL_checkAnalogOutputPort(port))
		throw port_exception(port, HALPort::AnalogOutput);
	handle = HAL_initializeAnalogOutputPort(port);
}
AnalogOutput::~AnalogOutput(){
	HAL_freeAnalogOutputPort(handle);
}

void AnalogOutput::setVoltage(float volts){
	HAL_setAnalogOutputVoltage(handle, volts);
}
void AnalogOutput::setValue(int value){
	HAL_setAnalogOutputValue(handle, value);
}

float AnalogOutput::getVoltage(){
	return HAL_getAnalogInputVoltage(handle);
}
int AnalogOutput::getValue(){
	return HAL_getAnalogInputValue(handle);
}

} /* namespace hal */

} /* namespace flashlib */

File: hal/java/src/edu/flash3388/flashlib/hal/AnalogOutput.java
package edu.flash3388.flashlib.hal;

public class AnalogOutput extends IOHandler{

	public AnalogOutput(int port){
		if(ANALOGJNI.checkPortAnalogOutput(port))
			throw new IllegalArgumentException("Unable to initialize analog output: "+port+" is not a valid analog output port");
		int handle = ANALOGJNI.initializeAnalogOutput(port);
		if(handle < 0)
			throw new IllegalStateException("Failed to initialize analog output port: "+port);
		setHandle(handle);
	}
	
	@Override
	public void free() {
		ANALOGJNI.freeAnalogOutput(getHandle());
	}

	public int getValue(){
		return ANALOGJNI.getAnalogValue(getHandle());
	}
	public double getVoltage(){
		return ANALOGJNI.getAnalogVoltage(getHandle());
	}
	
	public void setValue(int value){
		ANALOGJNI.setAnalogValue(getHandle(), value);
	}
	public void setVoltage(double volts){
		ANALOGJNI.setAnalogVoltage(getHandle(), (float)volts);
	}
}
